src/firebase-social-provider.js

```diff
@@ -0,0 +1,275 @@
+  // yet (i.e. they don't have their own inbox setup)
+  var toUserId = toClientId.substr(0, toClientId.indexOf('/'));
+  var toAgent = toClientId.substr(toClientId.indexOf('/'));
+  var friendInboxUrl =
+      this.baseUrl_ + '/' + this.networkName_ + ':' + toUserId + '/friends/' +
+      this.networkName_ + ':' + this.getUserId_() + '/inbox/' + toAgent;
+  var friendInboxRef = new Firebase(friendInboxUrl);
+  console.log('sending message to url: ' + friendInboxUrl);
+
+  // Send message in the format {fromAgent: message}
+  // This format is used so that we can monitor the 'child_added' event
+  // and get both the fromClientId and message in the same callback
+  var messageObj = {};
+  messageObj[this.loginState_.agent] = message;
+  friendInboxRef.push(messageObj);
+  return Promise.resolve();
+};
+
+FirebaseSocialProvider.prototype.logout = function() {
+  if (!this.loginState_) {
+    // User has already logged out.
+    return Promise.resolve();
+  }
+  this.setPresence_(false);
+  // TODO: we should disconnect all Firebase ref's here, so that we free up
+  // some of our Firebase quota.  However if we just call Firebase.goOffline(),
+  // additional attempts to login with the same webworker (freedom module)
+  // fail (authWithOAuthToken never returns).
+  this.initState_();
+  return Promise.resolve();
+};
+
+FirebaseSocialProvider.prototype.initState_ = function() {
+  this.baseUrl_ = null;
+  this.loginState_ = null;
+};
+
+// Friend should contain id and name fields
+FirebaseSocialProvider.prototype.addUserProfile_ = function(friend) {
+  if (this.loginState_.userProfiles[friend.userId]) {
+    console.warn('addUserProfile called for existing user ', friend);
+    return;
+  }
+
+  if (friend.userId == this.getUserId_()) {
+    // Add a user profile for ourself and just return.
+    // TODO: monitor for all clients except this one.
+    // TODO: refactor this so we don't have identical userProfile adding
+    // code twice.
+    this.loginState_.userProfiles[friend.userId] = {
+      userId: friend.userId,
+      name: friend.name || '',
+      lastUpdated: Date.now(),
+      url: friend.url || '',
+      imageData: friend.imageData || ''
+    };
+    this.dispatchEvent_(
+        'onUserProfile', this.loginState_.userProfiles[friend.userId]);
+    return;
+  }
+
+  // Ensure that a permanent friend object exists, with the users name.
+  // This must be present in order for other friends to properly detect our
+  // clients, based on the current Firebase rules configuration.
+  var myUrlForFriend =
+      this.baseUrl_ + '/' + this.networkName_ + ':' + this.getUserId_() +
+      '/friends/' + this.networkName_ + ':' + friend.userId;
+  var myRefForFriend = new Firebase(myUrlForFriend);
+  myRefForFriend.set({isFriend: true});
+
+  // Initialize an inbox, writable only by my friend, and unique to this
+  // agent (client).  This should be cleared when I disconnect.
+  var myInboxForFriendRef = new Firebase(
```

Comment:
This may be better as: `myRefForFriend.child('inbox/' + this.loginState_.agent);`